cpp/ql/src/Likely Bugs/Underspecified Functions/MistypedFunctionArguments.ql

```diff
@@ -0,0 +1,109 @@
+/**
+ * @name Call to a function with one or more incompatible arguments
+ * @description A call to a function with at least one argument whose type does
+ * not match the type of the corresponding function parameter.  This may indicate
+ * that the author is not familiar with the function being called.  Passing mistyped
+ * arguments on a stack may lead to unpredictable function behavior.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
+ * @id cpp/mistyped-function-arguments
+ * @tags correctness
+ *       maintainability
+ */
+
+import cpp
+
+pragma[inline]
+int sizeofInt() { result = any(IntType pt).getSize() }
+
+pragma[inline]
+predicate pointerArgTypeMayBeUsed(Type arg, Type parm) {
+  arg = parm
+  or
+  // arithmetic types
+  arg instanceof ArithmeticType and
+  parm instanceof ArithmeticType and
+  arg.getSize() = parm.getSize() and
+  (
+    (
+      arg instanceof IntegralType and
+      parm instanceof IntegralType
+    )
+    or
+    (
+      arg instanceof FloatingPointType and
+      parm instanceof FloatingPointType
+    )
+  )
+  or
+  // pointers to void are ok
+  arg instanceof VoidType
+  or
+  parm instanceof VoidType
+}
+
+pragma[inline]
+predicate argTypeMayBeUsed(Type arg, Type parm) {
+  arg = parm
+  or
+  // float will be promoted to double, and so it should correspond
+  // to the prototype
+  arg instanceof FloatType and parm instanceof DoubleType
+  or
+  // integral types are promoted "up to" (unsigned) int, but not long long.
+  arg instanceof IntegralType and
+  parm instanceof IntegralType and
+  arg.getSize() <= sizeofInt() and
+  parm.getSize() <= sizeofInt()
+  or
+  /*
+   *  // we allow interoperability between long long and pointer
+   *  arg.getSize() = parm.getSize() and
+   *  (
+   *    (arg instanceof IntegralType and parm instanceof PointerType)
+   *    or
+   *    (arg instanceof PointerType and parm instanceof IntegralType)
+   *  )
+   *  or
+   */
+
+  // pointers to compatible types
+  pointerArgTypeMayBeUsed(arg.(PointerType).getBaseType().getUnspecifiedType(),
+    parm.(PointerType).getBaseType().getUnspecifiedType())
+  or
+  pointerArgTypeMayBeUsed(arg.(PointerType).getBaseType().getUnspecifiedType(),
+    parm.(ArrayType).getBaseType().getUnspecifiedType())
+  or
+  pointerArgTypeMayBeUsed(arg.(ArrayType).getBaseType().getUnspecifiedType(),
+    parm.(PointerType).getBaseType().getUnspecifiedType())
+  or
+  pointerArgTypeMayBeUsed(arg.(ArrayType).getBaseType().getUnspecifiedType(),
+    parm.(ArrayType).getBaseType().getUnspecifiedType())
```

Comment:
I don't think it's possible for a fully-converted argument to be an `ArrayType`. It'll undergo an array-to-pointer conversion at some point. That means you can eliminate the last two of these cases.

Reply:
You're right.  I was thinking of C99/C11 fixed-length array parameters (e.g., <code>int arr[ static 10]</code>), but upon further research discovered that the <code>static 10</code> is merely a codegen hint for the compiler and that a raw pointer is still passed in.

Reply:
I thought about it some more, and now I think you do need the `arg.(ArrayType)` cases in order to handle the rarely-used "pointer to array" types (declared like `int (*parameterName)[2]`).

But hang on, where is the recursion that would enable this? If you allow passing a `int *` to a `unsigned int *` argument, shouldn't you also allow passing an `int **` to a `unsigned int **` argument? But recursive predicates can't be inline, so you can't support this without some serious restructuring for the sake of performance.

To start with, please make sure there is test coverage for "pointer to array" and "pointer to pointer" to compatible-but-not-identical base types.